File: ios/Plugin/IntercomBridgePlugin.m
#import <Foundation/Foundation.h>
#import <Capacitor/Capacitor.h>

// Define the plugin using the CAP_PLUGIN Macro, and
// each method the plugin supports using the CAP_PLUGIN_METHOD macro.
CAP_PLUGIN(IntercomBridgePlugin, "IntercomBridge",
           CAP_PLUGIN_METHOD(echo, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(registerUserWithUserId, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(registerUndefinedUser, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(logout, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(updateUser, CAPPluginReturnPromise);
           CAP_PLUGIN_METHOD(presentMessenger, CAPPluginReturnPromise);
)

File: ios/Plugin/IntercomBridgePlugin.swift
import Foundation
import Capacitor
import Intercom

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(IntercomBridgePlugin)
public class IntercomBridgePlugin: CAPPlugin {
    private let implementation = IntercomBridge()

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }
    
    @objc func registerUserWithUserId(_ call: CAPPluginCall) {
        let userId = call.getString("userId") ?? ""
        
        Intercom.registerUser(withUserId: userId)

        call.resolve([
            "value": "<#success#> registerUserWithUserId"
        ])
    }

    @objc func registerUndefinedUser(_ call: CAPPluginCall) {

        Intercom.registerUnidentifiedUser()

        call.resolve([
            "value": "<#success#> registerUndefinedUser"
        ])
    }
    
    @objc func logout(_ call: CAPPluginCall) {

        Intercom.logout()

        call.resolve([
            "value": "<#success#> logout"
        ])
    }
    
    @objc func updateUser(_ call: CAPPluginCall) {
        let name = call.getString("name") ?? ""
        let userId = call.getString("user_id") ?? ""
        
        let userAttributes = ICMUserAttributes()
        userAttributes.name = name
        if (userId != "") {
            userAttributes.userId = userId
        }
        
        Intercom.updateUser(userAttributes)

        call.resolve([
            "value": "<#success#> updateUser"
        ])
    }
    
    @objc func presentMessenger(_ call: CAPPluginCall) {
        Intercom.presentMessenger()
        
        call.resolve([
            "value": "<#success#> presentMessenger"
        ])
    }
}
